### PR TITLE
Enhance check functionality to support checking starting from a pageId

### PR DESCRIPTION
**Pull request by ahrtr**
Link to https://github.com/etcd-io/bbolt/issues/580

Only support specifying pageId in this PR. The reason of giving up supporting elementIdx is
- it complicates the overall design and implementation;
- the functionality of specifying elementIdx can also be achieved by only specifying pageId.

@ishan16696 @Elbehery @fuweid Please let me know if you have any comments on this PR.

@Elbehery Please feel free to cherry-pick the commit in this PR and add test cases in your PR.

### CODE REVIEW

File: tx_check.go [hunk 1]
```diff
@@ -56,19 +59,64 @@ func (tx *Tx) check(kvStringer KVStringer, ch chan error) {
...
+
+	switch {
+	case p.IsBranchPage():
+		for i := range p.BranchPageElements() {
+			elem := p.BranchPageElement(uint16(i))
+			tx.recursivelyCheckBucketInPage(elem.Pgid(), reachable, freed, kvStringer, ch)
+		}
+	case p.IsLeafPage():
+		for i := range p.LeafPageElements() {
+			elem := p.LeafPageElement(uint16(i))
```
Comment (fuweid) [inline tx_check.go:103]: IMO, we should try to get key/value here because the ksize/vsize might be wrong here

Comment (ahrtr) [reply tx_check.go:103]: It's a new checking, let's take care of it in a separate PR. Also it shouldn't be included in this method (each function/method should do only one thing); instead we can add a separate invariant checker in `checkInvariantProperties`, or gets it included in the existing `recursivelyCheckPageKeyOrder` (needs to change the method name if we follow this approach).

Comment (fuweid) [reply tx_check.go:103]: Sounds good.

**Review by fuweid — APPROVED** [review]
LGTM

File: tx_check.go [hunk 2]
```diff
@@ -27,11 +27,14 @@ func (tx *Tx) Check(options ...CheckOption) <-chan error {
...
 	ch := make(chan error)
-	go tx.check(chkConfig.kvStringer, ch)
+	go tx.check(chkConfig, ch)
 	return ch
 }
 
-func (tx *Tx) check(kvStringer KVStringer, ch chan error) {
+func (tx *Tx) check(cfg checkConfig, ch chan error) {
+	// Close the channel to signal completion.
+	defer close(ch)
```
Comment (ishan16696) [inline tx_check.go:36]: IMO, closing of channel should be the responsibility of caller.

Comment (ahrtr) [reply tx_check.go:36]: sort of agreeing on this, but not a big deal. 

Let's take care of in a followup PR with change in the method `Check` something like below,

```
go func() {
    defer close(ch)
    tx.check(chkConfig, ch)
}()
```

Comment (ahrtr) [reply tx_check.go:36]: Raised a separate PR https://github.com/etcd-io/bbolt/pull/666

**Review by ishan16696 — APPROVED** [review]
LGTM!!

**Comment by ahrtr** [comment]
Superseded by https://github.com/etcd-io/bbolt/pull/659

**Pull request closed without merging**